src/toxcore/tcp/handshake/packet.rs

```diff
@@ -0,0 +1,168 @@
+/*
+    Copyright (C) 2013 Tox project All Rights Reserved.
+    Copyright © 2017 Zetok Zalbavar <zexavexxe@gmail.com>
+    Copyright © 2017 Roman Proskuryakov <humbug@deeptown.org>
+
+    This file is part of Tox.
+
+    Tox is libre software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Tox is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*! Handshake packets to establish a confirmed connection via
+handshake using [`diagram`](https://zetok.github.io/tox-spec/#handshake-diagram)
+
+*/
+
+use toxcore::crypto_core::*;
+use toxcore::tcp::binary_io::*;
+
+use cookie_factory::*;
+
+/** The request of the client to create a TCP handshake.
+
+According to https://zetok.github.io/tox-spec/#handshake-request.
+
+Serialized form:
+
+Length  | Contents
+------- | --------
+`32`    | PK of the client
+`24`    | Nonce of the encrypted payload
+`72`    | Encrypted payload (plus MAC)
+
+*/
+
+#[derive(PartialEq, Debug, Clone)]
+pub struct ClientHandshake {
+    /// Client's Public Key
+    pub pk: PublicKey,
+    /// Nonce for the current encrypted payload
+    pub nonce: Nonce,
+    /// Encrypted payload according to
+    /// https://zetok.github.io/tox-spec/#handshake-request-packet-payload
+    pub payload: Vec<u8>
+}
+
+/// A serialized client handshake must be equal to 32 (PK) + 24 (nonce)
+/// + 72 (encrypted payload) bytes
+pub const CLIENT_HANDSHAKE_SIZE: usize = 128;
+
+impl FromBytes for ClientHandshake {
```

Comment:
There is from_bytes! macro that reduces a bit such definitions.

Reply:
I have got a different trait `FromBytes`, so your macro won't help me. This will be fixed when we find a solution that fit both for toxcore and toxcore/tcp

Reply:
But what the difference between your FromBytes and the existing one?

Reply:
Mine does not have bugs like `fn from_bytes(bytes: &[u8]) -> Option<Self> `